robotbase/src/main/java/com/team2357/frc2022/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,241 @@
+        public double m_turretMotorFF = 0;
+        public double m_turretMotorMaxOutput = 0;
+        public double m_turretMotorMinOutput = 0;
+        public double m_turretMotorMaxRPM = 0;
+        public double m_turretMotorMaxVel = 0;
+        public double m_turretMotorMinVel = 0;
+        public double m_turretMotorMaxAcc = 0;
+        public double m_turretMotorAllowedError = 0;
+
+        public double m_turretRotationsClockwiseSoftLimit = 0;
+        public double m_turretRotationsCounterClockwiseSoftLimit = 0;
+        public double m_rotationsPerDegree = 0;
+    }
+
+    public TurretSubsystem(CANSparkMax turretMotor) {
+        m_turretMotor = turretMotor;
+
+        m_arduinoHallEffectSensor = new ArduinoUSBController(Constants.ARDUINO.ARDUINO_SENSOR_DEVICE_NAME);
+
+        m_arduinoHallEffectSensor.start();
+
+        m_isZeroed = false;
+        resetHeading();
+    }
+
+    public void configure(Configuration config) {
+        m_config = config;
+
+        m_turretMotor.setIdleMode(m_config.turretMotorIdleMode);
+        m_turretMotor.setSmartCurrentLimit(m_config.m_turretMotorStallLimitAmps, m_config.m_turretMotorFreeLimitAmps);
+
+        m_pidController = m_turretMotor.getPIDController();
+
+        // PID coefficients
+        m_config.m_turretMotorP = 5e-5;
+        m_config.m_turretMotorI = 1e-6;
+        m_config.m_turretMotorD = 0;
+        m_config.m_turretMotorIZone = 0;
+        m_config.m_turretMotorFF = 0.000156;
+        m_config.m_turretMotorMinOutput = 1;
+        m_config.m_turretMotorMinOutput = -1;
+        m_config.m_turretMotorMaxRPM = 5700;
+
+        // Smart Motion Coefficients
+        m_config.m_turretMotorMaxVel = 2000; // rpm
+        m_config.m_turretMotorMaxAcc = 1500;
+
+        // set PID coefficients
+        m_pidController.setP(m_config.m_turretMotorP);
+        m_pidController.setI(m_config.m_turretMotorI);
+        m_pidController.setD(m_config.m_turretMotorD);
+        m_pidController.setIZone(m_config.m_turretMotorIZone);
+        m_pidController.setFF(m_config.m_turretMotorFF);
+        m_pidController.setOutputRange(m_config.m_turretMotorMinOutput, m_config.m_turretMotorMinOutput);
+
+        // Configure smart motion
+        int smartMotionSlot = 0;
+        m_pidController.setSmartMotionMaxVelocity(m_config.m_turretMotorMaxVel, smartMotionSlot);
+        m_pidController.setSmartMotionMinOutputVelocity(m_config.m_turretMotorMinVel, smartMotionSlot);
+        m_pidController.setSmartMotionMaxAccel(m_config.m_turretMotorMaxAcc, smartMotionSlot);
+        m_pidController.setSmartMotionAllowedClosedLoopError(m_config.m_turretMotorAllowedError, smartMotionSlot);
+    }
+
+    public void setTurretSpeed(double speed) {
+        m_turretMotor.set(speed);
+    }
+
+    public void resetHeading() {
+        m_turretMotor.getEncoder().setPosition(0);
+    }
+
+    public double getTurretRotations() {
+        return m_turretMotor.getEncoder().getPosition();
```

Comment:
This seems like an internal detail, also misnamed. The motor encoder position in clicks is probably meaningless outside of this subsystem if we're using degrees.

Reply:
This is not misnamed, SparkMax's use rotations as encoder position by default; however, your right that it will be meaningless once we change to degrees. The Neo's have a 3-phase hall-effect sensor and have 42 clicks per revolution by my understanding as well.